hooks/pre/30-el7_upgrade_postgresql.rb

```diff
@@ -10,9 +10,25 @@ def postgresql_12_upgrade
   ensure_packages(server_packages, 'installed')
 
   execute(%(scl enable rh-postgresql12 "PGSETUP_INITDB_OPTIONS='--lc-collate=#{collate} --lc-ctype=#{ctype} --locale=#{collate}' postgresql-setup --upgrade"))
-  ensure_packages(['postgresql', 'postgresql-server'], 'absent')
-  execute('rm -f /etc/systemd/system/postgresql.service')
-  ensure_packages(['rh-postgresql12-syspaths'], 'installed')
+
+  puppet_code = <<-EOF
+  package { 'postgresql-server':
+    ensure => absent,
+  } ->
+  package { 'postgresql':
+    ensure => absent,
+  } ->
+  file { '/etc/systemd/system/postgresql.service':
+    ensure => absent,
+  } ->
+  package { 'rh-postgresql12-syspaths':
+    ensure => installed,
+  }
+  EOF
+  failure_message = "Failed to ensure rh-postgresql12-syspaths is the system PostgreSQL"
+
+  logger.info("Ensuring rh-postgresql12-syspaths is the system PostgreSQL")
+  apply_puppet_code(puppet_code, failure_message)
```

Comment:
I get the concept and power of this code, the interface as a developer is just now much uglier :/ This is also a larger re-factoring of the change compared to:

```
 ensure_packages(['postgresql-server'], 'absent')
 ensure_packages(['postgresql'], 'absent')
```

I'll be curious to see how testing goes, but my vote is still the above.

Reply:
I considered the same. Given it needs to be cherry picked, I'm tempted to agree with you.